pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,262 @@
+
+type engine struct {
+	uuid [16]byte
+	conn pgxconn.PgxConn
+
+	stop chan struct{}
+
+	metrics sync.Map
+}
+
+type Telemetry interface {
+	Start()
+	// RegisterMetric registers a Prometheus Gauge or Counter metric for telemetry visitor.
+	// It must be called after creating the telemetry engine.
+	RegisterMetric(columnName string, gaugeOrCounterMetric prometheus.Metric) error
+	Stop()
+}
+
+func NewTelemetryEngine(conn pgxconn.PgxConn, uuid [16]byte) (Telemetry, error) {
+	// Warn the users about telemetry collection.
+	log.Warn("msg", "Promscale collects anonymous usage telemetry data for project improvements while being GDPR compliant. "+
+		"If you wish not to participate, please turn off sending telemetry in TimescaleDB. "+
+		"Details about the process can be found at https://docs.timescale.com/timescaledb/latest/how-to-guides/configuration/telemetry/#disabling-telemetry")
+	engine := &engine{
+		conn: conn,
+		uuid: uuid,
+	}
+	if err := engine.writeMetadata(); err != nil {
+		return nil, fmt.Errorf("writing metadata: %w", err)
+	}
+	return engine, nil
+}
+
+// writeMetadata writes Promscale and Tobs metadata.
+func (t *engine) writeMetadata() error {
+	promscale, err := promscaleMetadata()
+	if err != nil {
+		log.Debug("msg", "error fetching complete Promscale metadata", "error", err.Error())
+	}
+	promscale["promscale_exec_platform"] = ExecPlatform
+	if err := writeToTimescaleMetadataTable(t.conn, Stats(promscale)); err != nil {
+		return fmt.Errorf("writing metadata for promscale: %w", err)
+	}
+
+	tobs := tobsMetadata()
+	if len(tobs) > 0 {
+		if err := writeToTimescaleMetadataTable(t.conn, Stats(tobs)); err != nil {
+			return fmt.Errorf("writing metadata for tobs: %w", err)
+		}
+	}
+	return nil
+}
+
+const (
+	metadataUpdateWithExtension = "SELECT update_tsprom_metadata($1, $2, $3)"
+	metadataUpdateNoExtension   = "INSERT INTO _timescaledb_catalog.metadata(key, value, include_in_telemetry) VALUES ($1, $2, $3) ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value, include_in_telemetry = EXCLUDED.include_in_telemetry"
+)
+
+// writeToTimescaleMetadataTable syncs the metadata/stats with telemetry metadata table and returns the last error if any.
+func writeToTimescaleMetadataTable(conn pgxconn.PgxConn, m Stats) error {
+	update := func(queryFormat string) error {
+		batch := conn.NewBatch()
+		for key, metadata := range m {
+			safe := pgutf8str.Text{}
+			if err := safe.Set(metadata); err != nil {
+				return fmt.Errorf("setting in pgutf8 safe string: %w", err)
+			}
+			query := queryFormat
+			batch.Queue(query, key, safe, true)
+		}
+
+		results, err := conn.SendBatch(context.Background(), batch)
+		if err != nil {
+			return fmt.Errorf("error sending batch: %w", err)
+		}
+
+		rows, err := results.Query()
+		if err != nil {
+			// Metadata table not accessible. Try for func in Promscale extension.
+			return fmt.Errorf("error querying results: %w", err)
+		}
+		for rows.Next() {
+			if err != nil {
+				err = rows.Err()
+			}
```

Comment:
I don't this this if statement is correct unless I'm confused. What exactly is it supposed to do? It seems to me err will always be nil here.

Reply:
Ah, thanks for catching this. The if here is to catch the first most error and return, but I forgot to return the actual error.